File: app/src/main/java/net/feheren_fekete/applist/applistpage/DragGestureRecognizer.java
package net.feheren_fekete.applist.applistpage;

import android.graphics.PointF;
import androidx.core.view.MotionEventCompat;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

public class DragGestureRecognizer {

    public interface Callback {
        boolean canDrag(DragGestureRecognizer gestureRecognizer);
        boolean canStartDragging(DragGestureRecognizer gestureRecognizer);
        void onStartDragging(DragGestureRecognizer gestureRecognizer);
        void onDragging(DragGestureRecognizer gestureRecognizer);
        void onDrop(DragGestureRecognizer gestureRecognizer);
        void onStopDragging(DragGestureRecognizer gestureRecognizer);
    }

    private Callback mCallback;
    private ViewGroup mTouchOverlay;
    private View mDelegateView;
    private boolean mIsDelegateEnabled = true;
    private boolean mIsFingerDown;
    private boolean mIsDraggingItem;
    private PointF mFingerDownPos = new PointF();
    private MotionEvent mMotionEvent;

    public DragGestureRecognizer(Callback callback, ViewGroup touchOverlay, View delegateView) {
        mCallback = callback;
        mTouchOverlay = touchOverlay;
        mDelegateView = delegateView;

        mTouchOverlay.setOnTouchListener(mTouchOverlayListener);
    }

    public void setDelegateEnabled(boolean enabled) {
        mIsDelegateEnabled = enabled;
    }

    public MotionEvent getMotionEvent() {
        return mMotionEvent;
    }

    public PointF getFingerDownPos() {
        return mFingerDownPos;
    }

    private View.OnTouchListener mTouchOverlayListener = new View.OnTouchListener() {
        @Override
        public boolean onTouch(View v, MotionEvent event) {
            mMotionEvent = event;
            int action = MotionEventCompat.getActionMasked(event);
            switch (action) {
                case MotionEvent.ACTION_DOWN: {
                    mFingerDownPos.set(event.getRawX(), event.getRawY());
                    mIsFingerDown = true;
                    break;
                }
                case MotionEvent.ACTION_MOVE: {
                    if (mIsFingerDown) {
                        if (mCallback.canDrag(DragGestureRecognizer.this) && !mIsDraggingItem) {
                            // Tell the parent ViewGroup to let us handle
                            // events from now on. Otherwise the parent would steal them
                            // for handling it's own mechanisms.
                            // For example, a ViewPager would steel side-swipe events.
                            ((ViewGroup)v).requestDisallowInterceptTouchEvent(true);

                            if (mCallback.canStartDragging(DragGestureRecognizer.this)) {
                                mIsDraggingItem = true;
                                mCallback.onStartDragging(DragGestureRecognizer.this);
                            }
                        }
                        if (mIsDraggingItem) {
                            mCallback.onDragging(DragGestureRecognizer.this);
                        }
                    }
                    break;
                }
                case MotionEvent.ACTION_UP: {
                    if (mIsDraggingItem) {
                        mCallback.onDrop(DragGestureRecognizer.this);
                        mCallback.onStopDragging(DragGestureRecognizer.this);
                    }
                    mIsFingerDown = false;
                    mIsDraggingItem = false;
                    break;
                }
                case MotionEvent.ACTION_CANCEL: {
                    if (mIsDraggingItem) {
                        mCallback.onStopDragging(DragGestureRecognizer.this);
                    }
                    mIsFingerDown = false;
                    mIsDraggingItem = false;
                    break;
                }
            }
            if (!mIsDraggingItem) {
                if (mIsDelegateEnabled) {
                    mDelegateView.dispatchTouchEvent(event);
                }
            }
            return true;
        }
    };

}

File: app/src/main/java/net/feheren_fekete/applist/applistpage/DragGestureRecognizer.kt
package net.feheren_fekete.applist.applistpage

import android.graphics.PointF
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.core.view.MotionEventCompat

class DragGestureRecognizer(private val callback: Callback,
                            touchOverlay: ViewGroup,
                            private val delegateView: View) {

    private var isDelegateEnabled = true
    private var isFingerDown: Boolean = false
    private var isDraggingItem: Boolean = false

    val fingerDownPos = PointF()
    var motionEvent: MotionEvent? = null
        private set

    private val touchOverlayListener = View.OnTouchListener { v, event ->
        motionEvent = event
        val action = MotionEventCompat.getActionMasked(event)
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                fingerDownPos.set(event.rawX, event.rawY)
                isFingerDown = true
            }
            MotionEvent.ACTION_MOVE -> {
                if (isFingerDown) {
                    if (callback.canDrag(this@DragGestureRecognizer) && !isDraggingItem) {
                        // Tell the parent ViewGroup to let us handle
                        // events from now on. Otherwise the parent would steal them
                        // for handling it's own mechanisms.
                        // For example, a ViewPager would steel side-swipe events.
                        (v as ViewGroup).requestDisallowInterceptTouchEvent(true)

                        if (callback.canStartDragging(this@DragGestureRecognizer)) {
                            isDraggingItem = true
                            callback.onStartDragging(this@DragGestureRecognizer)
                        }
                    }
                    if (isDraggingItem) {
                        callback.onDragging(this@DragGestureRecognizer)
                    }
                }
            }
            MotionEvent.ACTION_UP -> {
                if (isDraggingItem) {
                    callback.onDrop(this@DragGestureRecognizer)
                    callback.onStopDragging(this@DragGestureRecognizer)
                }
                isFingerDown = false
                isDraggingItem = false
            }
            MotionEvent.ACTION_CANCEL -> {
                if (isDraggingItem) {
                    callback.onStopDragging(this@DragGestureRecognizer)
                }
                isFingerDown = false
                isDraggingItem = false
            }
        }
        if (!isDraggingItem) {
            if (isDelegateEnabled) {
                delegateView.dispatchTouchEvent(event)
            }
        }
        true
    }

    interface Callback {
        fun canDrag(gestureRecognizer: DragGestureRecognizer): Boolean
        fun canStartDragging(gestureRecognizer: DragGestureRecognizer): Boolean
        fun onStartDragging(gestureRecognizer: DragGestureRecognizer)
        fun onDragging(gestureRecognizer: DragGestureRecognizer)
        fun onDrop(gestureRecognizer: DragGestureRecognizer)
        fun onStopDragging(gestureRecognizer: DragGestureRecognizer)
    }

    init {
        touchOverlay.setOnTouchListener(touchOverlayListener)
    }

    fun setDelegateEnabled(enabled: Boolean) {
        isDelegateEnabled = enabled
    }

}
